airbyte-runner/src/main/java/io/airbyte/runner/RunnerApp.java

```diff
@@ -0,0 +1,171 @@
+        new AirbyteMessageTracker(),
+        new AirbyteMessageTracker());
+
+    LOGGER.info("Running replication worker...");
+    final Path jobRoot = WorkerUtils.getJobRoot(configs.getWorkspaceRoot(), jobRunConfig.getJobId(), jobRunConfig.getAttemptId());
+    final ReplicationOutput replicationOutput = replicationWorker.run(syncInput, jobRoot);
+
+    LOGGER.info("Sending output...");
+    // this uses stdout directly because it shouldn't have the logging related prefix
+    System.out.println(Jsons.serialize(replicationOutput));
+
+    LOGGER.info("Replication runner complete!");
+  }
+
+  public static void main(String[] args) throws Exception {
+    WorkerHeartbeatServer heartbeatServer = null;
+
+    try {
+      final String application = Files.readString(Path.of(ReplicationActivityImpl.INIT_FILE_APPLICATION));
+
+      final Map<String, String> envMap =
+          (Map<String, String>) Jsons.deserialize(Files.readString(Path.of(ReplicationActivityImpl.INIT_FILE_ENV_MAP)), Map.class);
+      final Configs configs = new EnvConfigs(envMap::get);
+
+      // set logging-related vars as properties so (at runtime) we can read these when processing
+      // log4j2.xml
+      setPropertyFromEnvVar(EnvConfigs.LOG_LEVEL);
+      setPropertyFromEnvVar(EnvConfigs.S3_PATH_STYLE_ACCESS);
+      System.setProperty(LogClientSingleton.S3_LOG_BUCKET, configs.getLogConfigs().getS3LogBucket());
+      System.setProperty(LogClientSingleton.S3_LOG_BUCKET_REGION, configs.getLogConfigs().getS3LogBucketRegion());
+      System.setProperty(LogClientSingleton.AWS_ACCESS_KEY_ID, configs.getLogConfigs().getAwsAccessKey());
+      System.setProperty(LogClientSingleton.AWS_SECRET_ACCESS_KEY, configs.getLogConfigs().getAwsSecretAccessKey());
+      System.setProperty(LogClientSingleton.S3_MINIO_ENDPOINT, configs.getLogConfigs().getS3MinioEndpoint());
+      System.setProperty(LogClientSingleton.GCP_STORAGE_BUCKET, configs.getLogConfigs().getGcpStorageBucket());
+      System.setProperty(LogClientSingleton.GOOGLE_APPLICATION_CREDENTIALS, configs.getLogConfigs().getGoogleApplicationCredentials());
+
+      final var logPath = LogClientSingleton.getInstance().getSchedulerLogsRoot(configs.getWorkspaceRoot());
+      LogClientSingleton.getInstance().setWorkspaceMdc(configs.getWorkerEnvironment(), configs.getLogConfigs(), logPath);
+
+      heartbeatServer = new WorkerHeartbeatServer(WorkerApp.KUBE_HEARTBEAT_PORT);
+      heartbeatServer.startBackground();
+
+      if (application.equals(ReplicationActivityImpl.REPLICATION)) {
+        replicationRunner(configs);
+      } else {
+        LOGGER.error("Runner failed", new IllegalStateException("Unexpected value: " + application));
+        System.exit(1);
+      }
+    } finally {
+      if (heartbeatServer != null) {
+        LOGGER.info("Shutting down heartbeat server...");
+        heartbeatServer.stop();
+      }
+    }
+
+    // required to kill s3 logger
```

Comment:
@davinchia I remember running into this when testing `KubePodProcess`. It's happening here too where we need to forcibly kill the process. Otherwise it just hangs on a background thread before closing...

Reply:
Gah. I don't remember why this was happening..